rel/marathon.cpp

```diff
@@ -0,0 +1,120 @@
+}
+
+// [Nambo] Transforms the current ball velocity relative to the goal, and stores it
+static void store_saved_vel() {
+    mkb::Ball& ball = mkb::balls[mkb::curr_player_idx];  // I think this makes ball.vel work
+
+    // Gives names to the item group index for the goal's IG, and the goal's index inside that IG
+    // This keeps the [] stuff inside the TF code short!
+    s16 goal_idx = mkb::mode_info.entered_goal_idx;
+    s16 goalIG_idx = mkb::mode_info.entered_goal_itemgroup_idx;
+
+    // MATRIX NOTES FROM COMPLEX (THE GOAT) BELOW vvv
+    // Build [world_from_goal]
+    // Build [world_from_itemgroup] * [itemgroup_from_goal]
+    // Build [world_from_itemgroup] * [translation(goal)] * [zrotation(goal)] * [yrotation(goal)] *
+    // [xrotation(goal)]
+
+    // Set up the TF matrix for Goal ---> World (Inverse of World ---> Goal)
+    mkb::mtxa_from_mtx(&mkb::itemgroups[goalIG_idx].transform);  // Build [world_from_itemgroup]
+    mkb::mtxa_translate(&mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].position);
+    mkb::mtxa_rotate_z(mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].rotation.z);
+    mkb::mtxa_rotate_y(mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].rotation.y);
+    mkb::mtxa_rotate_x(mkb::stagedef->coli_header_list[goalIG_idx].goal_list[goal_idx].rotation.x);
+    // Applies this TF ^^^
+    Vec entered_goal_vel_rt_world = {ball.vel.x, ball.vel.y, ball.vel.z};
+    Vec entered_goal_vel_rt_goal = {};
+    mkb::mtxa_rigid_inv_tf_vec(&entered_goal_vel_rt_world, &entered_goal_vel_rt_goal);
+    // We have now stored the desired veloocity vector as entered_goal_vel_rt_goal
+    s_saved_vel = entered_goal_vel_rt_goal;
+}
+
+void tick() {
+    // Mod is turned off
+    if (!pref::get_marathon() && s_state != MarathonState::WaitToStore) {
+        s_saved_vel = {};
+        s_state = MarathonState::WaitToStore;
+    }
+
+    // Mod is on
+    if (pref::get_marathon()) {
+        // If saved, apply the vel at 59.98 (1f after start)
+        if ((s_state == MarathonState::WaitToApply ||
+             s_state == MarathonState::WaitToApplyOrStore) &&
+            mkb::mode_info.stage_time_frames_remaining == mkb::mode_info.stage_time_limit - 1) {
+            apply_saved_vel();
+            s_state = MarathonState::WaitToApplyOrStore;
+        }
+
+        // If ready to store (the delay), store the velocity!
+        if (s_state == MarathonState::StoreNextFrame) {
+            store_saved_vel();
+            s_state = MarathonState::WaitToApply;
+        }
+
+        // At goal entry, get ready to store (This setup delays the action by 1 tick to avoid
+        // fricking with stuff)
+        if ((mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT ||
+             mkb::sub_mode == mkb::SMD_GAME_GOAL_MAIN) &&
+            (s_state == MarathonState::WaitToStore ||
+             s_state == MarathonState::WaitToApplyOrStore)) {
+            s_state = MarathonState::StoreNextFrame;
+        }
```

Comment:
This state machine is pretty confusing to me, I think you can reduce this to only three states. IMO it's helpful to think of the current value of `s_state` as the behavior you want to run this frame (and every frame you're in the state), and assigning `s_state` to mean, I want to be in the state starting _next frame_. You shouldn't transition multiple states on the same frame.

There's three states I see here, let me know if I'm missing something:

1. You are waiting to pass through the goal. You don't want to save the velocity on the frame you do end up passing through the goal, but rather on the next frame. Since the saving-ball-vel behavior is quite different than waiting-for-goal behavior (runs for many frames vs. one frame, checks a game condition vs. just writes to a variable), saving-ball-vel should be its own state.
2. You want to save the ball's velocity. This state is transitioned to from 1) upon goal entry, and immediately transitions to 3) after since its work is done. This state only runs for 1 frame.
3. You're waiting to apply the ball velocity. Since you want to apply the ball velocity on the same frame you detect the start of a level, you can apply it directly in this state - you don't really need a separate "apply vel" state. After this, you're back to the beginning of waiting to cross the goal, so you transition to 1) the next frame.

Each frame, you should only need to check if you're in one of these states (not if you're in one or the other), and only transition to one new state optionally. 

Hopefully this makes sense?